pkg/virt-launcher/virtwrap/api/converter.go

```diff
@@ -189,6 +189,29 @@ func SetDriverCacheMode(disk *Disk) error {
 	return nil
 }
 
+func SetFileDriverType(disk *Disk, diskInfoGetter func(string) (*containerdisk.DiskInfo, error)) error {
+	diskFileFormatUnset := disk.Type == "file" && disk.Source.File != "" && disk.Driver.Type == ""
+	if diskFileFormatUnset {
+		diskInfo, err := diskInfoGetter(disk.Source.File)
```

Comment:
We should always rely on a trusted source of information (other than the file itself) to determine if a file should be treated as a qcow2 image.  Several years ago there was a CVE where a malicious VM could write a fake qcow2 header and qcow2 metadata to the beginning of a raw disk image to trick the hypervisor into believing the file is qcow2.  This could lead to exposing sensitive data to the VM on subsequent reboots.  For this reason libvirt requires the format to be stated explicitly.  You will probably need to modify this PR so that the disk file format is exposed in the VM spec.  It's unfortunate that such a detail must be exposed to the end user but I do not see any way around it.

Reply:
> We should always rely on a trusted source of information (other than the file itself) to determine if a file should be treated as a qcow2 image. Several years ago there was a CVE where a malicious VM could write a fake qcow2 header and qcow2 metadata to the beginning of a raw disk image to trick the hypervisor into believing the file is qcow2. This could lead to exposing sensitive data to the VM on subsequent reboots. For this reason libvirt requires the format to be stated explicitly. You will probably need to modify this PR so that the disk file format is exposed in the VM spec. It's unfortunate that such a detail must be exposed to the end user but I do not see any way around it.

I think we should talk that through a little. It is out of question that qcow2 requires careful validation. But let's also keep in mind that we are now getting pretty close to the unprivileged pod scenario for running VMs. Even if one can escape it will not see anything which a user can't access anyway then.

I am not really convinced if it changes anything if a user specifies the disk type. I think this all boils down to no use questionable sources in general and it is less about the file format detection. Still i think that we can reach a pretty solid security barrier by properly validating the qcow and the encapsulation which pods give us. At the end I think that allowing people to pull and execute  container images per se is already as least as insecure as this scenario, where one would mostly just escape into the containers jail. vs. directly executing untrusted code in the pod.